### PR TITLE
[RLlib] Slate-Q tf implementation and tests/benchmarks.

### PR DESCRIPTION
**Pull request by sven1977**
Slate-Q tf implementation and tests/benchmarks.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by avnishn — APPROVED** [review]

File: rllib/agents/slateq/slateq.py [hunk 1]
```diff
@@ -107,19 +105,28 @@
...
     "adam_epsilon": 1e-8,
     # If not None, clip gradients during optimization at this value
     "grad_clip": 40,
     # How many steps of the model to sample before learning starts.
-    "learning_starts": 1000,
+    "learning_starts": 20000,
     # Update the replay buffer with this many samples at once. Note that
     # this setting applies per-worker if num_workers > 1.
-    "rollout_fragment_length": 1000,
+    "rollout_fragment_length": 4,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq.py:129]: Sorry, forgot to explain: To more closely match Dopamine's implementation.

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 2]
```diff
@@ -129,7 +129,32 @@ def build_slateq_losses(
     # 'batch_indexes']
     learning_strategy = policy.config["slateq_strategy"]
 
-    if learning_strategy == "SARSA":
+    # Myopic agent: Don't care about value of next state.
+    # Acts only based off immediate reward.
+    if learning_strategy == "MYOP":
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_torch_policy.py:134]: I just moved the if-blocks around for more clarity: Found it weird that the "main" SlateQ implementation ("QL") was buried somewhere. This should not have any effect.

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    score_no_click = tf.fill([tf.shape(user_obs)[0], 1], no_click_score)
+    # Concatenate click and no-click scores.
+    all_scores = tf.concat([scores_per_candidate, score_no_click], axis=1)
+
+    # Logits: Softmax to yield probabilities.
+    if multinomial_logits:
+        all_scores = tf.nn.softmax(all_scores)
+    # Multinomial proportional model: Shift to `[0.0,..[`.
+    else:
+        all_scores = all_scores - min_normalizer
```
Comment (gjoliver) [inline rllib/agents/slateq/slateq_tf_policy.py:262]: hmmm, this is also kind of very recsim specific.
I think at the very least, we need to document multinomial_logits=False and min_normalizer=-1.0 above well.

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    Args:
+        policy: The Policy object.
+        obs_space: The Policy's observation space.
+        action_space: The Policy's action space.
+        config: The Policy's config.
+    """
+    TargetNetworkMixin.__init__(policy, config)
+
+
+def rmsprop_optimizer(
```
Comment (gjoliver) [inline rllib/agents/slateq/slateq_tf_policy.py:320]: we don't have an existing util file that defines all the optimizers?

File: rllib/examples/env/recommender_system_envs_with_recsim.py [hunk 5]
```diff
@@ -94,7 +92,7 @@ def __init__(self, doc_id, features, video_length=None, quality=None):
 
 
 def iev_document_sampler_creator(env_ctx):
-    return iev.IEvVideoSampler(doc_ctor=SingleClusterIEvVideo, seed=env_ctx["seed"])
+    return iev.UtilityModelVideoSampler(doc_ctor=iev.IEvVideo, seed=env_ctx["seed"])
```
Comment (gjoliver) [inline rllib/examples/env/recommender_system_envs_with_recsim.py:95]: if you meant to revert these back, you can delete the SingleClusterIEvVideo class above.
I'd be interested in seeing the obs vectors, and understand what our documents look like ...

File: rllib/utils/exploration/slate_epsilon_greedy.py [hunk 6]
```diff
@@ -12,6 +12,64 @@
...
+            else explore,
+            true_fn=(lambda: tf.where(chose_random, random_actions, exploit_action)),
+            false_fn=lambda: exploit_action,
+        )
+
+        if self.framework in ["tf2", "tfe"] and not self.policy_config["eager_tracing"]:
+            self.last_timestep = timestep
+            return action, action_logp
+        else:
+            assign_op = tf1.assign(self.last_timestep, tf.cast(timestep, tf.int64))
```
Comment (gjoliver) [inline rllib/utils/exploration/slate_epsilon_greedy.py:69]: sorry, why do we need to keep track of and update timestep here?
will this timestep get restored if job crashes etc?

**Review by gjoliver — COMMENTED** [review]
sorry about the delay. a few questions, thanks.

File: rllib/agents/slateq/slateq.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq.py:131]: I'll add a comment. Reason I changed to 4 was to make it closer to Dopamine's implementation and to our DQN.
In my opinion, it doesn't make sense to wait for an episode to finish before sending it to the buffer. It's 1-step Q learning anyways.

File: rllib/agents/bandit/bandit.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
     "rollout_fragment_length": 1,
     "train_batch_size": 1,
 
-    # Bandits cant afford to do one timestep per iteration as it is extremely
+    # Bandits can't afford to do one timestep per iteration as it is extremely
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandit/bandit.py:23]: It's the original (AWS) comment. Good point, should clarify.
The problem with 1 is that then we would report (in tune) every single timestep (b/c every iteration is just 1 timestep in the num_workers=0, num_envs_per_worker=0, train_batch_size=0 default setup), which would be madness.

File: rllib/agents/slateq/slateq.py [hunk 8]
```diff
@@ -139,9 +148,10 @@
     # Learning method used by the slateq policy. Choose from: RANDOM,
     # MYOP (myopic), SARSA, QL (Q-Learning),
     "slateq_strategy": "QL",
-    # Only relevant for `slateq_strategy="QL"`:
+
+    # TODO: Unify torch and tf configs.
     # Use double_q correction to avoid overestimation of target Q-values.
-    "double_q": True,
+    "double_q": True,  # Only relevant for `slateq_strategy="QL"` and framework="torch".
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq.py:154]: We don't currently. And I'm not sure anymore we should: The target term is generated from `target_net([obs])`, not `target_net([next_obs])` it seems (still not 100% certain this is correct, but RecSim's agent does it this way). We should try it though.

File: rllib/agents/slateq/slateq_tf_model.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        Extra model kwargs:
+            fcnet_hiddens_per_candidate: List of layer-sizes for each(!) of the
+                candidate documents.
+        """
+        super(SlateQTFModel, self).__init__(
+            obs_space, action_space, None, model_config, name
+        )
+
+        self.orig_obs_space = obs_space.original_space
+        self.embedding_size = self.orig_obs_space["doc"]["0"].shape[0]
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_model.py:40]: I think it's ok that it's a specialized algo (on certain env types with certain obs-space restrictions). We should add a `validate_spaces` function to the slate-q policy.

File: rllib/agents/slateq/slateq_tf_model.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        for i in range(self.num_candidates):
+            doc = self.docs_in[
+                :, self.embedding_size * i : self.embedding_size * (i + 1)
+            ]
+            out = tf.keras.layers.concatenate([self.user_in, doc], axis=1)
+            for h in fcnet_hiddens_per_candidate:
+                out = tf.keras.layers.Dense(h, activation=tf.nn.relu)(out)
+            q_value = tf.keras.layers.Dense(1, name=f"q_value_{i}")(out)
+            q_outs.append(q_value)
+        q_outs = tf.concat(q_outs, axis=1)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_model.py:64]: When we do the slate combinatorics, we throw out duplicate entries via tf.boolean_mask.

File: rllib/agents/slateq/slateq_tf_policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_policy.py:320]: Nope :D

Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_policy.py:320]: We should make optimizers configurable as we do exploration/(soon) buffers/inputs:

```
config:
  optimizer_config:
    type: Adam|RMSprop|...
    lr: ...
    epsilon: ...
```

File: rllib/examples/env/recommender_system_envs_with_recsim.py (hunk 5 above)
Comment (sven1977) [reply rllib/examples/env/recommender_system_envs_with_recsim.py:95]: Just wanted to make things comparable again between RecSim's agent and ours.

File: rllib/utils/exploration/slate_epsilon_greedy.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/exploration/slate_epsilon_greedy.py:69]: Yes, it should. It's part of the exploration component's state.

**Pull request merged**